changes/tasks.py

```diff
@@ -1468,6 +1468,82 @@ def run(self, vumi_message_id: str, event_type: str, **kwargs) -> None:
 process_whatsapp_system_event = ProcessWhatsAppSystemEvent()
 
 
+class ProcessWhatsAppTimeoutSystemEvent(Task):
+    """
+    Notifies a user that a message we send them on Whatsapp has not been delivered
+    within 7 days, she will be sent an outbound
+    message informing them and give the option to switch to SMS.
+    """
+
+    name = "ndoh_hub.changes.tasks.process_whatsapp_timeout_system_event"
+
+    def return_date_time(self):
+        return datetime.utcnow()
+
+    def send_outbound(self, user_id, source_id, identity_uuid):
+        identity = is_client.get_identity(identity_uuid)
+        # Transform to django language code
+        for item in identity["results"]:
+            details = item["details"]
+            if "lang_code" in details:
+                language = details["lang_code"].lower().replace("_", "-")
+                with translation.override(language):
+                    text = translation.ugettext(
+                        "We see that your MomConnect WhatsApp messages are not being "
+                        "delivered. If you would like to receive your messages over "
+                        "SMS, reply ‘SMS’."
+                    )
+
+                utils.ms_client.create_outbound(
+                    {
+                        "to_identity": identity_uuid,
+                        "content": text,
+                        "channel": "JUNE_TEXT",
+                        "metadata": {},
+                    }
+                )
+
+    def handle_undelivered(self, user_id, source_id, identity_uuid):
+        identity = is_client.get_identity(identity_uuid)
+        for item in identity["results"]:
+            details = item["details"]
+            if "timeout_timestamp" not in details:
+                date = self.return_date_time()
+                details["timeout_timestamp"] = date.timestamp()
+                is_client.update_identity({"id": item["id"], "details": details})
+                self.send_outbound(user_id, source_id, identity_uuid)
+
+            else:
+                d1 = datetime.fromtimestamp(details["timeout_timestamp"])
+                d2 = self.return_date_time()
+                delta = (d2 - d1).days
+                if delta >= settings.WHATSAPP_EXPIRY_SMS_BOUNCE_DAYS:
+                    details["timeout_timestamp"] = d2.timestamp()
+                    is_client.update_identity({"id": item["id"], "details": details})
+                    self.send_outbound(user_id, source_id, identity_uuid)
```

Comment:
Maybe we can just send the whole identity here, so that we don't have to do the second HTTP request in `send_outbound`

Reply:
By send the whole identity, do you mean send it in the update_identityline?  is the a method to do both an update and a send

Reply:
I meant that instead of doing a `self.send_outbound(user_id, source_id, identity_uuid)`, and then performing another request to the identity store in `send_outbound`, we should rather do a `self.send_outbound(user_id, source_id, identity)`, and then the `send_outbound` function has all the identity details, and doesn't need to do a request to the identity store.

Reply:
oh I see what you meant, that's brilliant thank you